### PR TITLE
Try tests on GitHub Actions with three MPI ranks

### PR DESCRIPTION
**Pull request by jhale**
Can keep the others on two for coverage? To discuss.

### CODE REVIEW

**Comment by jhale** [comment]
```
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 3
slots that were requested by the application:

  ctest

Either request fewer procs for your application, or make more slots
available for use.

A "slot" is the PRRTE term for an allocatable unit where we can
launch a process.  The number of slots available are defined by the
environment in which PRRTE processes are run:

  1. Hostfile, via "slots=N" clauses (N defaults to number of
     processor cores if not provided)
  2. The --host command line parameter, via a ":N" suffix on the
     hostname (N defaults to 1 if not provided)
  3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
  4. If none of a hostfile, the --host command line parameter, or an
     RM is present, PRRTE defaults to the number of processor cores

In all the above cases, if you want PRRTE to default to the number
of hardware threads instead of the number of processor cores, use the
--use-hwthread-cpus option.

Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
number of available slots when deciding the number of processes to
launch.
--------------------------------------------------------------------------
```

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

Strange as the public runners all have at least 3 processors.

**Comment by jorgensd** [comment]
> ```
> --------------------------------------------------------------------------
> There are not enough slots available in the system to satisfy the 3
> slots that were requested by the application:
> 
>   ctest
> 
> Either request fewer procs for your application, or make more slots
> available for use.
> 
> A "slot" is the PRRTE term for an allocatable unit where we can
> launch a process.  The number of slots available are defined by the
> environment in which PRRTE processes are run:
> 
>   1. Hostfile, via "slots=N" clauses (N defaults to number of
>      processor cores if not provided)
>   2. The --host command line parameter, via a ":N" suffix on the
>      hostname (N defaults to 1 if not provided)
>   3. Resource manager (e.g., SLURM, PBS/Torque, LSF, etc.)
>   4. If none of a hostfile, the --host command line parameter, or an
>      RM is present, PRRTE defaults to the number of processor cores
> 
> In all the above cases, if you want PRRTE to default to the number
> of hardware threads instead of the number of processor cores, use the
> --use-hwthread-cpus option.
> 
> Alternatively, you can use the --map-by :OVERSUBSCRIBE option to ignore the
> number of available slots when deciding the number of processes to
> launch.
> --------------------------------------------------------------------------
> ```
> 
> https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories
> 
> Strange as the public runners all have at least 3 processors.

Ubuntu runders only have two processors. Mac tynnere have more

**Comment by jhale** [comment]
Not according to githubs own documentation?

**Comment by jorgensd** [comment]
> Not according to githubs own documentation?

are we reading the same documentation?
IMG_8445

**Comment by jhale** [comment]
No, there are two sections, one for public repositories and one for private. You are reading the latter.

**Comment by jorgensd** [comment]
> No, there are two sections, one for public repositories and one for private. You are reading the latter.

Interesting! That is a new development? 

I had the same oversubscription issue when using openmpi with adios4dolfinx a while back. I haven't seen it with mpich before.

File: .github/workflows/ccpp.yml [hunk 1]
```diff
@@ -107,7 +108,7 @@ jobs:
       - name: Run C++ unit tests (MPI)
         run: |
           cd build/test
-          mpiexec -np 2 ctest -V --output-on-failure -R unittests
+          mpirun -n 3 --map-by :OVERSUBSCRIBE ctest -V --output-on-failure -R unittests
```
Comment (garth-wells) [inline .github/workflows/ccpp.yml:111]: Don't use MPI implementation specific args - we can't then switch MPI version. Use environment variables (we already use some OpenMPI env variables on the CI).

See https://github.com/open-mpi/ompi/issues/8955#issuecomment-846272943, https://github.com/open-mpi/ompi/issues/8955#issuecomment-1295456308 and https://github.com/mpi4py/mpi4py-testing/blob/1dcad02e04f8c65c3e31416dcb86a7b8c1affcb8/.github/workflows/openmpi.yml#L98.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**